libdnf-plugins/actions/actions.cpp

```diff
@@ -0,0 +1,725 @@
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "utils/bgettext/bgettext-mark-domain.h"
+
+#include <fmt/format.h>
+#include <libdnf/base/base.hpp>
+#include <libdnf/base/transaction.hpp>
+#include <libdnf/common/exception.hpp>
+#include <libdnf/rpm/package_query.hpp>
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include <cerrno>
+#include <cstring>
+#include <filesystem>
+#include <fstream>
+#include <optional>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+using namespace libdnf::plugin;
+
+namespace {
+
+constexpr const char * PLUGIN_NAME = "actions";
+constexpr PluginVersion PLUGIN_VERSION{0, 1, 0};
+
+constexpr const char * attrs[]{"author.name", "author.email", "description", nullptr};
+constexpr const char * attrs_value[]{"Jaroslav Rohel", "jrohel@redhat.com", "Actions Plugin."};
+
+// Represents one line in action file
+struct Action {
+    std::filesystem::path file_path;
+    int line_number;
+    std::string pkg_filter;
+    enum class Direction { IN, OUT, ALL } direction;
+    std::string command;
+    std::vector<std::string> args;
+};
+
+
+// Represents one command to run
+struct CommandToRun {
+    [[nodiscard]] bool operator<(const CommandToRun & other) const noexcept;
+    std::string command;
+    std::vector<std::string> args;
+};
+
+
+class Actions : public IPlugin {
+public:
+    Actions(libdnf::Base & base, libdnf::ConfigParser &) : base(base) {}
+    virtual ~Actions() = default;
+
+    APIVersion get_api_version() const noexcept override { return PLUGIN_API_VERSION; }
+
+    const char * get_name() const noexcept override { return PLUGIN_NAME; }
+
+    PluginVersion get_version() const noexcept override { return PLUGIN_VERSION; }
+
+    const char * const * get_attributes() const noexcept override { return attrs; }
+
+    const char * get_attribute(const char * attribute) const noexcept override {
+        for (size_t i = 0; attrs[i]; ++i) {
+            if (std::strcmp(attribute, attrs[i]) == 0) {
+                return attrs_value[i];
+            }
+        }
+        return nullptr;
+    }
+
+    void init() override { parse_action_files(); }
+
+    void pre_base_setup() override { setup(pre_base_setup_actions); }
+
+    void post_base_setup() override { setup(post_base_setup_actions); }
+
+    void pre_transaction(const libdnf::base::Transaction & transaction) override {
+        transaction_impl(transaction, pre_trans_actions);
+    }
+
+    void post_transaction(const libdnf::base::Transaction & transaction) override {
+        transaction_impl(transaction, post_trans_actions);
+    }
+
+private:
+    void parse_action_files();
+    void setup(const std::vector<Action> & trans_actions);
+    void transaction_impl(const libdnf::base::Transaction & transaction, const std::vector<Action> & trans_actions);
```

Comment:
How about naming the two above methods `on_base_setup()` and `on_transaction()` for better consistency?

Reply:
OK